web3/_utils/contracts.py

```diff
@@ -294,120 +222,102 @@ def prepare_transaction(
 ) -> HexStr:
+    info_abi: ABIElement
     if fn_identifier is FallbackFn:
-        fn_abi, fn_selector, fn_arguments = get_fallback_function_info(
-            contract_abi, fn_abi
+        info_abi, info_selector, info_arguments = get_fallback_function_info(
+            contract_abi, cast(ABIFallback, fn_abi)
         )
     elif fn_identifier is ReceiveFn:
-        fn_abi, fn_selector, fn_arguments = get_receive_function_info(
-            contract_abi, fn_abi
+        info_abi, info_selector, info_arguments = get_receive_function_info(
+            contract_abi, cast(ABIReceive, fn_abi)
         )
-    elif is_text(fn_identifier):
-        fn_abi, fn_selector, fn_arguments = get_function_info(
-            # type ignored b/c fn_id here is always str b/c FallbackFn is handled above
-            fn_identifier,  # type: ignore
-            w3.codec,
+    elif isinstance(fn_identifier, str):
+        fn_info = get_abi_element_info(
             contract_abi,
-            fn_abi,
-            args,
-            kwargs,
+            fn_identifier,
+            *args,
+            abi_codec=w3.codec,
+            **kwargs,
         )
+        info_abi = fn_info["abi"]
+        info_selector = fn_info["selector"]
+        info_arguments = fn_info["arguments"]
     else:
         raise Web3TypeError("Unsupported function identifier")
 
-    return add_0x_prefix(encode_abi(w3, fn_abi, fn_arguments, fn_selector))
+    return add_0x_prefix(encode_abi(w3, info_abi, info_arguments, info_selector))
 
 
 def decode_transaction_data(
     fn_abi: ABIFunction,
     data: HexStr,
     normalizers: Sequence[Callable[[TypeStr, Any], Tuple[TypeStr, Any]]] = None,
 ) -> Dict[str, Any]:
-    data = HexBytes(data)
+    data_bytes = HexBytes(data)
     types = get_abi_input_types(fn_abi)
     abi_codec = ABICodec(default_registry)
-    decoded = abi_codec.decode(types, HexBytes(data[4:]))
+    decoded = abi_codec.decode(types, data_bytes[4:])
     if normalizers:
         decoded = map_abi_data(normalizers, types, decoded)
     return named_tree(fn_abi["inputs"], decoded)
 
 
-def get_fallback_function_info(
-    contract_abi: Optional[ABI] = None, fn_abi: Optional[ABIFunction] = None
-) -> Tuple[ABIFunction, HexStr, Tuple[Any, ...]]:
+def get_constructor_function_info(
+    contract_abi: Optional[ABI] = None, fn_abi: Optional[ABIConstructor] = None
+) -> Tuple[ABIConstructor, HexStr, Tuple[Any, ...]]:
     if fn_abi is None:
-        fn_abi = get_fallback_func_abi(contract_abi)
+        fn_abi = get_constructor_function_abi(contract_abi)
     fn_selector = encode_hex(b"")
     fn_arguments: Tuple[Any, ...] = tuple()
     return fn_abi, fn_selector, fn_arguments
 
 
-def get_receive_function_info(
-    contract_abi: Optional[ABI] = None, fn_abi: Optional[ABIFunction] = None
-) -> Tuple[ABIFunction, HexStr, Tuple[Any, ...]]:
+def get_fallback_function_info(
+    contract_abi: Optional[ABI] = None, fn_abi: Optional[ABIFallback] = None
```

Comment:
nit

```suggestion
    contract_abi: Optional[ABI] = None, fallback_abi: Optional[ABIFallback] = None
```

Reply:
I dont think these are all that useful since `get_abi_element` achieves this.

I could see these being convenient for users by returning the appropriate typed ABI element though. Do you think it's worth refactoring these functions to use `get_abi_element`?

Reply:
This still lives under `_utils`. Let's keep the scope of this PR more narrow and focus on breaking *eth-utils* related changes for now. If we want to add convenient methods for users, we can do so as features later on. Whatever makes sense for internal use, let's make those changes here.

Reply:
Oh that makes things easier then, these are not used anywhere and can be removed. 👍